jina/drivers/search.py

```diff
@@ -46,9 +46,9 @@ def _apply_all(self, docs: Iterable['jina_pb2.Document'], *args, **kwargs):
             else:
                 miss_idx.append(idx)
 
-        # delete non-existed matches in reverse
-        for j in reversed(miss_idx):
-            del docs[j]
+        # # delete non-existed matches in reverse
```

Comment:
Can u explain here how we identified the need to avoid this deletion, it was quite a case, not obvious at all

Reply:
We had a query which was split into sentences (e.g. documents with granularity 1). For some reason they completely disappeared from the data stream. After going backwards, we found out, that something around the `KVSearchDriver` caused. Until ultimately we saw that driver deleting the chunks, since they have no entry in the `BinaryPbIndexer` (they are part of the query). And the driver things, whatever has no value in my database, is not worth returning.

Reply:
this logic has to be reserved. this handles the miss docs in sharding.

Reply:
so @maximilianwerk we need to adapt somehow the logic of traversing to adapt then to this without removing this, maybe we need to recover the `recur_on` parameter?

Reply:
@hanxiao it disappeared because the way we provide now `granularity_range` and `adjacency_range` is limiting. It came from an unwanted loop on `chunks` before `matches`. So we will need to tackle this.

Reply:
@maximilianwerk It would be great to add a unit test so that we ensure everyone is on the same page.

Reply:
Yes adding tests was anyhow on my agenda.

Reply:
it is mostly caused by the limitation on how we traverse. In one case we are traversing on chunks before traversing in matches because we `set granularity_range: [1, 1] and adjacency_range: [0 , 1]` to go for the KVSearchDrivrer. So it is a "bug more or less" it first does an iteration on chunks and does not find anything because is a query chunk, therefore it deletes the chunk from the query. So later when we traverse on matches it has no chunk to go down to